### DISCUSSION TITLE
Avoid calculating material properties need only for Jacobian

### DISCUSSION THREAD

**Discussion by dzheng26** [idaholab/moose#27892] [Q&A General] [answered]
### Check these boxes if you have followed the posting rules.

- [X] Q&A General is the most appropriate section for my question
- [X] I have consulted the posting Guidelines on the Discussions front page
- [X] I have searched the Discussions forum and my question has not been asked before
- [X] I have searched the MOOSE website and the documentation does not answer my question
- [X] I have formatted my post following the posting guidelines (screenshots as a last resort, triple back quotes around pasted text)

### Question

For one nonlinear iteration, the method ```computeProperties``` of the material classes is called at least 2 times, once for calculating the residual and once for calculating the Jacobian.

Some properties generated by ```computeProperties``` are actually for calculating Jacobian only, for example the ```pk1_jacobian``` in solid mechanics. When ```computeProperties``` is called for calculating the residual, the calculation for ```pk1_jacobian``` is not really needed. For complicate material model, the calculation for ```pk1_jacobian``` maybe relatively heavy and slows down the simulation.

So my question is:
Is there a way to know in ```computeProperies``` if it is used for calculating the residual or the jacobian, or both (for Executioner/residual_and_jacobian_together=1)? If it is known, user can skip the properties which are used for the Jacobian only when ```computeProperties``` is called not for calculation the Jacobian.

**Comment by GiudGiud** [comment, marked as answer]
Hello

Yes there is a way. You can check the problem with "currentlyComputingJacobian"
```
_fe_problem.currentlyComputingJacobian()
```
returns a boolean.

Guillaume

**Reply by dzheng26** [reply]
Thanks for the quick answer. I will try it.
